packages/react/src/components/TextField/index.tsx

```diff
@@ -1,143 +1,122 @@
+
     const handleChange = useCallback(
-      (value: string) => {
-        const count = countCodePointsInString(value)
+      (e: React.ChangeEvent<HTMLInputElement>) => {
+        const value = e.target.value
+        const count = getCount(value)
         if (maxLength !== undefined && count > maxLength) {
           return
         }
-        if (nonControlled) {
-          setCount(count)
-        }
+        setCount(count)
         onChange?.(value)
       },
-      [maxLength, nonControlled, onChange]
+      [getCount, maxLength, onChange]
     )
 
     useEffect(() => {
-      setCount(countCodePointsInString(props.value ?? ''))
-    }, [props.value])
-
-    const {
-      inputProps: ariaInputProps,
-      labelProps,
-      descriptionProps,
-      errorMessageProps,
-    } = useTextField(
-      {
-        inputElementType: 'input',
-        isDisabled: disabled,
-        isRequired: required,
-        isReadOnly: readOnly,
-        validationState: invalid ? 'invalid' : 'valid',
-        description: !invalid && assistiveText,
-        errorMessage: invalid && assistiveText,
-        onChange: handleChange,
-        ...props,
-      },
-      ariaRef
-    )
+      setCount(getCount(value ?? ''))
+    }, [getCount, value])
 
     const containerRef = useRef(null)
 
-    useFocusWithClick(containerRef, ariaRef)
+    useFocusWithClick(containerRef, inputRef)
+
+    const inputId = useId(props.id)
+    const describedbyId = useId()
+    const labelledbyId = useId()
 
-    const inputProps = mergeProps(restProps, ariaInputProps)
+    const showAssistiveText =
+      assistiveText != null && assistiveText.length !== 0
 
     return (
       <TextFieldRoot className={className} isDisabled={disabled}>
         <FieldLabel
+          htmlFor={inputId}
+          id={labelledbyId}
           label={label}
-          requiredText={requiredText}
           required={required}
+          requiredText={requiredText}
           subLabel={subLabel}
-          {...labelProps}
           {...(!showLabel ? visuallyHiddenProps : {})}
         />
         <StyledInputContainer
-          ref={containerRef}
-          invalid={invalid}
           aria-disabled={disabled === true ? true : undefined}
           hasPrefix={prefix != null}
           hasSuffix={suffix != null || showCount}
+          invalid={props.invalid === true}
+          ref={containerRef}
         >
           {prefix && <PrefixContainer>{prefix}</PrefixContainer>}
           <StyledInput
-            ref={mergeRefs(forwardRef, ariaRef)}
-            invalid={invalid}
-            {...inputProps}
+            aria-describedby={showAssistiveText ? describedbyId : undefined}
+            aria-invalid={props.invalid}
+            aria-labelledby={labelledbyId}
+            id={inputId}
+            invalid={props.invalid === true}
+            maxLength={maxLength}
+            onChange={handleChange}
+            prefix={props.htmlPrefix}
```

Comment:
It may not be used in the real world, but there is a definition for it in [react/types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L2924).

Naming it `rdfaPrefix` might be more accurate.

Reply:
I was not aware of the existence of `prefix` in `react/types`.  Thanks for that.

I guess `rdfaPredix` is easier for users to understand than `htmlPrefix`, since `react/types` also says RDFa Attributes.